### PR TITLE
feat: add `auth` field to `RPCEndpointConfig`

### LINKED ISSUE #8276: feat(config): add support for authentication in `[rpc_endpoints]`
### Component

Forge

### Describe the feature you would like

https://github.com/foundry-rs/foundry/blob/f9674c3ea80ec9cd92e5f96a75542b3a26fa4752/crates/config/src/endpoints.rs#L215-L215

we already support settings like cups and retries:

https://github.com/foundry-rs/foundry/blob/f9674c3ea80ec9cd92e5f96a75542b3a26fa4752/crates/config/src/endpoints.rs#L219-L228

we can easily integrate authentication as well,

this should probably be something like

```rust
auth = "bearer|basic:${BEARER}"
```

## TODO
* support parsing auth field
* integrate in provider setup

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by caiquejjx**
## Motivation
Closes #8276 

## Solution
Adds the auth field to RPCEndpointConfig

### CODE REVIEW

File: crates/config/src/lib.rs [hunk 1]
```diff
@@ -886,7 +886,7 @@ impl Config {
                         if self.offline {
                             return Err(SolcError::msg(format!(
                                 "can't install missing solc {version} in offline mode"
-                            )))
```
Comment (zerosnacks) [inline crates/config/src/lib.rs:889]: Nit: please undo the inserted `;`

See: https://github.com/foundry-rs/foundry/blob/master/CONTRIBUTING.md#resolving-an-issue for the linter settings (we are using `nightly` fmt and `nightly` clippy)

**Comment by zerosnacks** [comment]
Thanks for your PR! Small note on the unintended formatting diff

**Comment by caiquejjx** [comment]
done, thanks @zerosnacks

**Review by mattsse — CHANGES_REQUESTED** [review]
great start,

we can expect that this will be used via some env var or similar so we should add support for resolving env vars

see for ref

https://github.com/foundry-rs/foundry/blob/c99854277c346fa6de7a8f9837230b36fd85850e/crates/config/src/endpoints.rs#L117-L131

and

https://github.com/foundry-rs/foundry/blob/c99854277c346fa6de7a8f9837230b36fd85850e/crates/config/src/endpoints.rs#L150-L160

**Comment by caiquejjx** [comment]
@mattsse I have implemented the auth as an enum similar to the `RpcEndpoint`, lmk you thoughts

File: crates/config/src/endpoints.rs [hunk 2]
```diff
@@ -210,6 +210,58 @@ impl From<RpcEndpoint> for RpcEndpointConfig {
     }
 }
 
+/// The auth token to be used for RPC endpoints
+/// It works in the same way as the `RpcEndpoint` type, where it can be a raw string or a reference
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum RpcAuth {
+    Token(String),
```
Comment (mattsse) [inline crates/config/src/endpoints.rs:217]:
```suggestion
    Raw(String),
```

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, would love to see some additional tests like

https://github.com/foundry-rs/foundry/blob/cce36f85c80946bd4a1868411aa99eda879a0e43/crates/config/src/lib.rs#L3369-L3369

**Comment by caiquejjx** [comment]
@mattsse I added a test for checking the resolving of the auth and I also added the `auths` field to the `ResolvedRpcEndpoint` as a way of resolving the auth, its probably not the most optimal solution but is the one that requires less changes on existing code. I can also create a new type for the `config.resolve` result where I can add the auth for an endpoint but let me know what you think

**Review by mattsse — APPROVED** [review]
cool, nice work.

next up we'd find a way to integrate this properly

I think the entrypoint for this would be:

https://github.com/foundry-rs/foundry/blob/master/crates/cli/src/utils/mod.rs#L92-L92

**Comment by mattsse** [comment]
failing tests unrelated

**Pull request merged**